sdks/python/apache_beam/testing/metric_result_matchers.py

```diff
@@ -0,0 +1,261 @@
+                'original_name' : 'ToIsmRecordForMultimap-out0-ElementCount',
+                'output_user_name' : 'ToIsmRecordForMultimap-out0'
+            }),
+            metric_result_matchers.is_committed_counter(42),
+            metric_result_matchers.is_attempted_counter(42)
+        ),
+        all_of(
+            metric_result_matchers.has_name('MeanByteCount'),
+            metric_result_matchers.has_labels({
+                'original_name' : 'Read-out0-MeanByteCount',
+                'output_user_name' : 'GroupByKey/Read-out0'
+            }),
+            metric_result_matchers.is_committed_counter(31),
+            metric_result_matchers.is_attempted_counter(31)
+        ),
+        all_of(
+            metric_result_matchers.has_name('ExecutionTime_ProcessElement'),
+            metric_result_matchers.has_step_name('write/Write/Write'),
+            metric_result_matchers.is_committed_counter(1000),
+            metric_result_matchers.is_attempted_counter(1000)
+        ),
+    ]
+    errors = metric_result_matchers.verify_all(
+        result.metrics().all_metrics(), matchers)
+    self.assertFalse(errors, errors)
+"""
+
+from __future__ import absolute_import
+
+from apache_beam.typehints import with_input_types
+from apache_beam.metrics.cells import DistributionResult
+
+from hamcrest import equal_to
+from hamcrest.core.matcher import Matcher
+from hamcrest.core.base_matcher import BaseMatcher
+from hamcrest.core.core.allof import all_of
+from hamcrest.core import string_description
+
+import typing
+
+
+def _matcher_or_equal_to(value_or_matcher):
+  """Pass-thru for matchers, and wraps value inputs in an equal_to matcher."""
+  if value_or_matcher is None:
+    return None
+  if isinstance(value_or_matcher, Matcher):
+    return value_or_matcher
+  return equal_to(value_or_matcher)
+
+
+class MetricResultMatcher(BaseMatcher):
+  """A PyHamcrest matcher that validates counter MetricResults."""
+
+  @with_input_types(
+      namespace=typing.Tuple[str, Matcher],
+      name=typing.Tuple[str, Matcher],
+      step=typing.Tuple[str, Matcher],
+      label_key=typing.Tuple[str, Matcher],
+      label_value=typing.Tuple[str, Matcher],
+      attempted_value=typing.Tuple[int, Matcher],
+      committed_value=typing.Tuple[int, DistributionResult, Matcher])
+  def __init__(self, namespace=None, name=None, step=None, label_key=None,
+               label_value=None, attempted_value=None, committed_value=None):
+    self.namespace = _matcher_or_equal_to(namespace)
+    self.name = _matcher_or_equal_to(name)
+    self.step = _matcher_or_equal_to(step)
+    self.label_key = _matcher_or_equal_to(label_key)
+    self.label_value = _matcher_or_equal_to(label_value)
+    self.attempted_value = _matcher_or_equal_to(attempted_value)
+    self.committed_value = _matcher_or_equal_to(committed_value)
+
+  def _matches(self, metric_result):
+    if self.namespace is not None and not self.namespace.matches(
```

Comment:
any significance to the `is not None` here vs just truthiness check on the next couple string fields below?

Reply:
this is none of my business, but I was in the neighborhood. I imagine an empty string is a valid namespace, so there's that.